### PR TITLE
[SYCL] Remove free function queries call detection from FE

### PR DESCRIPTION
**Pull request by alexbatashev**
Removing free function call detection from clang front-end dramatically improves compile times of device-side code (up to 3 times in some cases).

This has the following implications:
1. Range rounding is no longer disabled based on free-function queries. Users are encouraged to disable range rounding manually.
2. Free function queries are no longer supported on host device.

### CODE REVIEW

File: clang/include/clang/Sema/Sema.h [hunk 1]
```diff
@@ -383,15 +383,6 @@ class SYCLIntegrationHeader {
     KernelParamDesc() = default;
   };
 
-  // there are four free functions the kernel may call (this_id, this_item,
-  // this_nd_item, this_group)
-  struct KernelCallsSYCLFreeFunction {
```
Comment (elizabethandrews) [inline clang/include/clang/Sema/Sema.h:388]: Lines 362-367 can also be deleted

```
  /// Note which free functions (this_id, this_item, etc) are called within the
  /// kernel
  void setCallsThisId(bool B);
  void setCallsThisItem(bool B);
  void setCallsThisNDItem(bool B);
  void setCallsThisGroup(bool B);
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -3143,60 +3143,6 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
...
-
-    while (!WorkList.empty()) {
-      const FunctionDecl *FD = WorkList.back().first;
-      WorkList.pop_back();
-      if (!Visited.insert(FD).second)
-        continue; // We've already seen this Decl
-
-      // Check whether this call is to free functions (sycl::this_item(),
-      // this_id, etc.).
-      if (Util::isSyclFunction(FD, "this_id")) {
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:3169]: I think you can also delete `isSyclFunction `declaration and definition in this file. I believe it is only used for this purpose

File: clang/test/SemaSYCL/Inputs/CL/sycl/detail/kernel_desc.hpp [hunk 3]
```diff
@@ -43,8 +43,6 @@ __SYCL_INLINE_NAMESPACE(cl) {
     }
     static constexpr const char *getName() { return ""; }
     static constexpr bool isESIMD() { return 0; }
-    static constexpr bool callsThisItem() { return false; }
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/Inputs/CL/sycl/detail/kernel_desc.hpp:46]: Can you also please check  `clang/test/CodeGenSYCL/Inputs/sycl.hpp`. I'm not sure but there might be code to be removed there as well

**Review by elizabethandrews — COMMENTED** [review]
Can you please explain why removing FE detection changes this behavior you mentioned in description?

`Free function queries are no longer supported on host device`.

File: sycl/include/CL/sycl/group.hpp [hunk 4]
```diff
@@ -422,19 +422,15 @@ template <int Dimensions = 1> class group {
...
 template <int Dims>
 __SYCL_DEPRECATED("use sycl::ext::oneapi::experimental::this_group() instead")
 group<Dims> this_group() {
 #ifdef __SYCL_DEVICE_ONLY__
   return detail::Builder::getElement(detail::declptr<group<Dims>>());
 #else
-  return detail::store_group<Dims>(nullptr);
+  throw sycl::exception(
+      sycl::make_error_code(sycl::errc::feature_not_supported),
+      "Free function calls are not supported on host device");
```
Comment (smanna12) [inline sycl/include/CL/sycl/group.hpp:433]: Why does this error happen?

File: sycl/include/CL/sycl/group.hpp [hunk 5]
```diff
@@ -445,7 +441,9 @@ template <int Dims> group<Dims> this_group() {
 #ifdef __SYCL_DEVICE_ONLY__
   return sycl::detail::Builder::getElement(detail::declptr<group<Dims>>());
 #else
-  return sycl::detail::store_group<Dims>(nullptr);
+  throw sycl::exception(
+      sycl::make_error_code(sycl::errc::feature_not_supported),
+      "Free function calls are not supported on host device");
```
Comment (smanna12) [inline sycl/include/CL/sycl/group.hpp:446]: Same question here?

**Comment by alexbatashev** [comment]
> Can you please explain why removing FE detection changes this behavior you mentioned in description?
> > Free function queries are no longer supported on host device

@elizabethandrews @smanna12 these queries are not native to host device and runtime relies on integration header to make sure we store necessary info. This makes this feature usable only if integration header is present. A runtime-only implementation is possible, but it would be inefficient and hurt runtime performance. And since host device is rarely used, and the feature can not be implemented efficiently, it is much easier to simply drop this feature support for host device. This aligns with C++ policy "do not pay for what you don't use".

**Comment by smanna12** [comment]
> > Can you please explain why removing FE detection changes this behavior you mentioned in description?
> > > Free function queries are no longer supported on host device
> 
> @elizabethandrews @smanna12 these queries are not native to host device and runtime relies on integration header to make sure we store necessary info. This makes this feature usable only if integration header is present. A runtime-only implementation is possible, but it would be inefficient and hurt runtime performance. And since host device is rarely used, and the feature can not be implemented efficiently, it is much easier to simply drop this feature support for host device. This aligns with C++ policy "do not pay for what you don't use".

Thanks for the explanation.

**Comment by alexbatashev** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/654

**Review by elizabethandrews — APPROVED** [review]
FE changes LGTM

**Review by smanna12 — APPROVED** [review]
FE changes look good to me.

**Review by sergey-semenov — APPROVED** [review]
Runtime part LGTM

**Comment by alexbatashev** [comment]
@psalz this commit should also help with performance of CTS compilation

**Pull request merged**